src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DictionaryStringHelper.cs

```diff
@@ -33,5 +33,32 @@ internal static string ReadAnyAsnString(this AsnReader tavReader)
                     throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
             }
         }
+
+        internal static string ReadAnyAsnString(ref this AsnValueReader tavReader)
+        {
+            Asn1Tag tag = tavReader.PeekTag();
+
+            if (tag.TagClass != TagClass.Universal)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            switch ((UniversalTagNumber)tag.TagValue)
+            {
+                case UniversalTagNumber.BMPString:
+                case UniversalTagNumber.IA5String:
+                case UniversalTagNumber.NumericString:
+                case UniversalTagNumber.PrintableString:
+                case UniversalTagNumber.UTF8String:
+                case UniversalTagNumber.T61String:
+                    // .NET's string comparisons start by checking the length, so a trailing
```

Comment:
What about null characters embedded in the middle of the string?

Reply:
Those would compare as different, which is fine.  The problem that the preexisting version of this code was working around is that Windows has, over the years, produced strings where they encoded the trailing \0 and they were matching in native code, but not in managed.